Mathlib/CategoryTheory/Category/RelCat.lean

```diff
@@ -35,4 +39,128 @@ instance rel : LargeCategory RelCat where
   comp f g x z := ∃ y, f x y ∧ g y z
 #align category_theory.rel CategoryTheory.rel
 
+@[ext] theorem rel_ext {X Y : RelCat} (f g : X ⟶ Y) (h : ∀ a b, f a b ↔ g a b) : f = g :=
+  funext₂ (fun a b => propext (h a b))
+
+namespace RelCat.Hom
+
+protected theorem rel_id (X : RelCat) : 𝟙 X = (· = ·) := rfl
+
+protected theorem rel_comp {X Y Z : RelCat} (f : X ⟶ Y) (g : Y ⟶ Z) : f ≫ g = Rel.comp f g := rfl
+
+@[simp] theorem rel_id_apply₂ (X : RelCat) (x y : X) : (𝟙 X) x y ↔ x = y := by
+  rw [RelCat.Hom.rel_id]
+
+@[simp] theorem rel_comp_apply₂ {X Y Z : RelCat} (f : X ⟶ Y) (g : Y ⟶ Z) (x : X) (z : Z) :
+    (f ≫ g) x z ↔ ∃ y, f x y ∧ g y z := by rfl
+
+end RelCat.Hom
+
+/-- The essentially surjective faithful embedding
+from the category of types and functions into the category of types and relations. -/
+def graphFunctor : Type u ⥤ RelCat.{u} where
+  obj X := X
+  map f := f.graph
+
+@[simp] theorem graphFunctor_map {X Y : Type u} (f : X ⟶ Y) (x : X) (y : Y) :
+    graphFunctor.map f x y ↔ f x = y := f.graph_def x y
+
+instance graphFunctor_faithful : Faithful graphFunctor where
+  map_injective h := Function.graph_injective h
+
+instance graphFunctor_essSurj : EssSurj graphFunctor :=
+    graphFunctor.essSurj_of_surj Function.surjective_id
+
+open Classical
+/-- A relation is an isomorphism in `RelCat` iff it is the image of an isomorphism in
+`Type`. -/
+theorem rel_iso_iff {X Y : RelCat} (r : X ⟶ Y) :
+    IsIso (C := RelCat) r ↔ ∃ f : (Iso (C := Type) X Y), graphFunctor.map f.hom = r := by
+  constructor
+  · intro h
+    have h1 := congr_fun₂ h.hom_inv_id
+    have h2 := congr_fun₂ h.inv_hom_id
+    simp only [RelCat.Hom.rel_comp_apply₂, RelCat.Hom.rel_id_apply₂, eq_iff_iff] at h1 h2
+    obtain ⟨f, hf⟩ := axiomOfChoice (fun a => (h1 a a).mpr rfl)
+    obtain ⟨g, hg⟩ := axiomOfChoice (fun a => (h2 a a).mpr rfl)
+    suffices hif : IsIso (C := Type) f by
+      use asIso f
+      ext x y
+      simp only [asIso_hom, graphFunctor_map]
+      constructor
+      · rintro rfl
+        exact (hf x).1
+      · intro hr
+        specialize h2 (f x) y
+        rw [← h2]
+        use x, (hf x).2, hr
+    use g
+    constructor
+    · ext x
+      apply (h1 _ _).mp
+      use f x, (hg _).2, (hf _).2
+    · ext y
+      apply (h2 _ _).mp
+      use g y, (hf (g y)).2, (hg y).2
+  · rintro ⟨f, rfl⟩
+    apply graphFunctor.map_isIso
+
+section Opposite
+open Opposite
+
+/-- The argument-swap isomorphism from `rel` to its opposite. -/
+def relRelOp : RelCat ⥤ RelCatᵒᵖ where
```

Comment:
The name `RelCat.opFunctor` (and similarly bellow `RelCat.unopFunctor`) might be better.

Reply:
thanks!